p2p/pex/file.go

```diff
@@ -16,16 +16,18 @@ type addrBookJSON struct {
 }
 
 func (a *addrBook) saveToFile(filePath string) {
-	a.Logger.Info("Saving AddrBook to file", "size", a.Size())
-
 	a.mtx.Lock()
 	defer a.mtx.Unlock()
-	// Compile Addrs
-	addrs := []*knownAddress{}
+
+	bookSize := a.size()
+	a.Logger.Info("Saving AddrBook to file", "size", bookSize)
+
+	addrs := make([]*knownAddress, bookSize)
```

Comment:
Do we know if using `i` here is significantly better than using `make([]*knownAddress, 0, bookSize)` and `append`?

Reply:
It's faster

Reply:
The previous is more obviously correct.  With custom counters I need to think harder to know that it's correct.  For one, I need to look back up at the implementation of .size() in order to know that there is sufficient capacity in addrs, otherwise it will panic.  And, as this code mutates, it's easy to add a continue statement in the wrong place (e.g. before the i++) (this is what I mean by increased surface area for bugs).

The previous code was more clean and beautiful, and more obviously panic-free.

There's another way to make it more performant, while retaining the obvious correctness and simplicity/less-surface-area.

```
        addrs := make([]*knownAddress, 0, bookSize)
	for _, ka := range a.addrLookup {
		addrs = append(addrs, ka)
	}
```

Reply:
> The previous code was more clean and beautiful, and more obviously panic-free.

Agree

> I also need to look back up at the implementation of .size() in order to know that there is sufficient capacity in addrs, otherwise it will panic.

One might argue that if it panics, then our software is incorrect and it's easy to bring this fact to our attention. Otherwise, software will be running, but it may be incorrect and nobody will notice it (until somebody opens an issue "Why is my addrbook contains only 50% of addresses").

Reply:
Sure, but it would still be better to avoid the usage of .size() altogether and just use len(addrLookup) instead.  I'm not advocating for covering bugs, but rather making code less brittle to changes.  By not relying on .size() when we're going to iterate over a thing that supports len() anyways, we also get to change the definition of size, etc.

Reply:
Will do.